Multipart reader regression in go 1.17

### Checklist

- [X] This is a bug report, not a question. Ask questions on [discuss.ipfs.io](https://discuss.ipfs.io).
- [X] I have searched on the [issue tracker](https://github.com/ipfs/go-ipfs/issues?q=is%3Aissue) for my bug.
- [X] I am running the latest [go-ipfs version](https://dist.ipfs.io/#go-ipfs) or have an issue updating.

### Installation method

built from source

### Version

_No response_

### Config

_No response_

### Description

See the test failure in https://github.com/ipfs/go-ipfs-files/pull/40:

```
--- FAIL: TestMultipartFiles (0.00s)
    helpers_test.go:122: [2] found end of directory, expected files.Event{kind:0, name:"nested", value:"some content"}
FAIL
FAIL	github.com/ipfs/go-ipfs-files	0.007s
FAIL
```

This isn't a flake so _something_ is going wrong. I'm not sure if this affects production, but we should figure it out before releasing.

--- a/docs/examples/go-ipfs-as-a-library/go.mod
+++ b/docs/examples/go-ipfs-as-a-library/go.mod
@@ -5,7 +5,7 @@ go 1.16
 require (
 	github.com/ipfs/go-ipfs v0.9.1
 	github.com/ipfs/go-ipfs-config v0.16.0
-	github.com/ipfs/go-ipfs-files v0.0.8
+	github.com/ipfs/go-ipfs-files v0.0.9
 	github.com/ipfs/interface-go-ipfs-core v0.5.1
 	github.com/libp2p/go-libp2p-core v0.9.0
 	github.com/multiformats/go-multiaddr v0.4.0

--- a/go.mod
+++ b/go.mod
@@ -33,7 +33,7 @@ require (
 	github.com/ipfs/go-ipfs-config v0.16.0
 	github.com/ipfs/go-ipfs-exchange-interface v0.0.1
 	github.com/ipfs/go-ipfs-exchange-offline v0.0.1
-	github.com/ipfs/go-ipfs-files v0.0.8
+	github.com/ipfs/go-ipfs-files v0.0.9
 	github.com/ipfs/go-ipfs-keystore v0.0.2
 	github.com/ipfs/go-ipfs-pinner v0.1.2
 	github.com/ipfs/go-ipfs-posinfo v0.0.1

